libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -22,12 +22,78 @@
 #endif // no system header
 
 #include <cuda/std/__atomic/functions/cuda_ptx_generated.h>
+#include <cuda/std/__type_traits/conditional.h>
+#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__type_traits/is_scalar.h>
 #include <cuda/std/cstdint>
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
 #if defined(_CCCL_CUDA_COMPILER)
 
+template <typename _Tp,
+          typename _Sco,
+          __enable_if_t<!is_scalar<_Tp>::value && (sizeof(_Tp) == 4 || sizeof(_Tp) == 8), int> = 0>
+_CCCL_DEVICE bool __atomic_compare_exchange_cuda(
+  _Tp volatile* __ptr,
+  _Tp* __expected,
+  const _Tp __desired,
+  bool __weak,
+  int __success_memorder,
+  int __failure_memorder,
+  _Sco)
+{
+  using __proxy_t = _If<sizeof(_Tp) == 4, uint32_t, uint64_t>;
+  __proxy_t __old = 0;
+  __proxy_t __new = 0;
+  memcpy(&__old, __expected, sizeof(__proxy_t));
+  memcpy(&__new, &__desired, sizeof(__proxy_t));
+  bool __result = __atomic_compare_exchange_cuda(
+    reinterpret_cast<volatile __proxy_t*>(__ptr), &__old, __new, __weak, __success_memorder, __failure_memorder, _Sco{});
```

Comment:
These casts are illegal.

I can see a way around it, though - have the underlying integer versions of these functions, and the PTX wrappers, take void pointers for the first arguments. Since it is used only in the inline PTX, it doesn't actually need to carry the type information with it, so that should be fine.

Reply:
I believe this is now resolved.